Server Scopes

sjpadgett opened the pull request:
- consolidated all scope handling to ScopeRepository
- fixed up fhir metadata and moved scope qualify for smart well known to scope repository

<!--
(Thanks for sending a pull request!
-->

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->
Fixes #4121 

#### Short description of what this resolves:
Closing in on final scope and claim handling

sjpadgett commented:
@bradymiller @adunsulag 
Not sure I can explain all I did. Basically consolidated scopes in one place(scope repository).
The goal was to be able to validate and only use scopes appropriate to client application role and context.
So i've added all potential scopes and the permissions we want to allow an app to do. Then depending on what fhir resources we have currently active(routes) and their implementation(read or write) I validate against our core scope defaults arriving at a current list of valid scopes.

I did it this way so when adding new api resources or permission, we shouldn't have to edit or hardcoded core claims and scopes.

I did this in rather a hurry because I keep trying to get to buttoning down registration and writing manual register. So change or move how you see fit where hopefully i've given a framework how this could work.

adunsulag commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -132,4 +73,323 @@ public function finalizeScopes(
+     *
+     * @return array
+     */
+    public function fhirScopes(): array
```
So eventually your thought is we will put these into the database and let the site administrator turn this on or off right?

Just trying to make sure I understand that these are the **permitted** scopes that we are allowing FHIR authors to make calls against versus what the system technologically supports which are defined in the REST api.  I think I'm following everything here but I just want to double check.

sjpadgett replied:
Yep, they are a list of all current or future supported scopes. Where we only publish scopes the api supports at the time validated against this list.
Hopefully when we add a new Fhir resource to our routes then scopes will already be in place

sjpadgett replied:
> Only thing I'm not entirely clear on is if our scopes listed in the oauth2/provider/.well-known/discovery.php need to match what we list in our SMART well-known.

hmm, I need to think about that for a minute. I'm leaning towards you are prob right and should, especially if we offer the api:fhir scope. What about standard apis? Should we limit them by scopes as we do for Fhir?

adunsulag replied:
It'd be great to keep them all consistent, but I don't fully understand all the uses cases for the standard api like you and Brady do, so I would think that'd be y'alls call.

sjpadgett replied:
It may be what @bradymiller was speaking about with a custom scope for native however, since we are putting in place the permissions then perhaps we should scope standard as well. Makes no sense to limit Fhir access and not standard too.

sjpadgett replied:
I'm thinking we should preface our scopes with `fhir:user/Patient.read` and `standard:user/patient.read`

adunsulag replied:
I'm not following.  ONC testing won't send any **fhir:** prefix. Would you stripe out the prefix when comparing against a SMART app registration or authorization request?

sjpadgett replied:
I could but maybe instead just preface for our standard api thus leaving fhir alone. OpenEMR I think is in a spot by offering two api sets. It is just very handy for vendor apps to be able to call both or many vendors will write apps using only our standard api.
Bringing into question; do we register client apps as only Fhir or standard, public/user/system facing on a per client basis or allow one client to support several apps. A phone call please so I can explain better:)

bradymiller replied:
we can schedule a call today for 11 :)

sjpadgett replied:
works for me. can you do @adunsulag

adunsulag reviewed (commented):
The changes look good to me.  Only thing I'm not entirely clear on is if our scopes listed in the oauth2/provider/.well-known/discovery.php need to match what we list in our SMART well-known.

sjpadgett commented on oauth2/provider/.well-known/discovery.php:
```diff
@@ -1,23 +1,24 @@
 $discovery = <<<TEMPLATE
 {
-"issuer": "$authServer->authIssueFullUrl",
+"issuer": "$base_url",
```
I moved it to oauth2/authorize.php so I wouldn't need to declare the authserver global in discovery.

sjpadgett commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -132,4 +73,323 @@ public function finalizeScopes(
+        if ($role === 'user') {
+            $standard = array_merge($this->fhirRequiredSmartScopes(), $permitted);
+        }
+        if ($role === 'patient') {
```
right.

sjpadgett commented:
Here are our standard apis scopes gen'ed from routes
```
 "patient/drug.read",
    "patient/facility.write",
    "patient/immunization.read",
    "patient/insurance_company.write",
    "patient/patient.write",
    "patient/practitioner.write",
    "patient/prescription.read",
    "patient/procedure.read",
    "system/drug.read",
    "system/facility.write",
    "system/immunization.read",
    "system/insurance_company.write",
    "system/patient.write",
    "system/practitioner.write",
    "system/prescription.read",
    "system/procedure.read",
    "user/drug.read",
    "user/facility.write",
    "user/immunization.read",
    "user/insurance_company.write",
    "user/patient.write",
    "user/practitioner.write",
    "user/prescription.read",
    "user/procedure.read"
```
@bradymiller I guess Stephen is busy. We can talk if you want or wait for Stephen..

stephenwaite commented:
feel free to call in rn @sjpadgett , @bradymiller and @adunsulag

sjpadgett commented:
lol, I just realized today is Sat and this was a normal Sat call! duh

bradymiller commented:
lol, feels like a sunday :)

sjpadgett commented:
sorry guys, got dragged away, maybe an hour more. you're right @bradymiller standard scopes were not right woops..

sjpadgett commented:
Well here's discovery result
```
{
  "issuer": "https://opensourcedemr.us/oauth2/default",
  "authorization_endpoint": "https://opensourcedemr.us/oauth2/default/authorize",
  "token_endpoint": "https://opensourcedemr.us/oauth2/default/token",
  "jwks_uri": "https://opensourcedemr.us/oauth2/default/jwk",
  "userinfo_endpoint": "https://opensourcedemr.us/oauth2/default/userinfo",
  "registration_endpoint": "https://opensourcedemr.us/oauth2/default/registration",
  "end_session_endpoint": "https://opensourcedemr.us/oauth2/default/logout",
  "introspection_endpoint": "https://opensourcedemr.us/oauth2/default/introspect",
  "scopes_supported": [
    "openid",
    "fhirUser",
    "online_access",
    "offline_access",
    "launch",
    "launch/patient",
    "api:oemr",
    "api:fhir",
    "api:port",
    "api:pofh",
    "profile",
    "name",
    "address",
    "given_name",
    "family_name",
    "nickname",
    "phone",
    "phone_verified",
    "email",
    "email_verified",
    "patient/AllergyIntolerance.read",
    "patient/CareTeam.read",
    "patient/Condition.read",
    "patient/Encounter.read",
    "patient/Immunization.read",
    "patient/Location.read",
    "patient/MedicationRequest.read",
    "patient/Observation.read",
    "patient/Organization.read",
    "patient/Organization.write",
    "patient/Patient.read",
    "patient/Patient.write",
    "patient/Procedure.read",
    "system/AllergyIntolerance.read",
    "system/CareTeam.read",
    "system/Condition.read",
    "system/Encounter.read",
    "system/Immunization.read",
    "system/Location.read",
    "system/MedicationRequest.read",
    "system/Observation.read",
    "system/Organization.read",
    "system/Organization.write",
    "system/Patient.read",
    "system/Patient.write",
    "system/Practitioner.read",
    "system/Practitioner.write",
    "system/Procedure.read",
    "user/AllergyIntolerance.read",
    "user/CareTeam.read",
    "user/Condition.read",
    "user/Encounter.read",
    "user/Immunization.read",
    "user/Location.read",
    "user/MedicationRequest.read",
    "user/Observation.read",
    "user/Organization.read",
    "user/Organization.write",
    "user/Patient.read",
    "user/Patient.write",
    "user/Practitioner.read",
    "user/Practitioner.write",
    "user/Procedure.read",
    "patient/allergy.read",
    "patient/appointment.read",
    "patient/drug.read",
    "patient/facility.read",
    "patient/facility.write",
    "patient/immunization.read",
    "patient/insurance_company.read",
    "patient/insurance_company.write",
    "patient/list.read",
    "patient/medical_problem.read",
    "patient/patient.read",
    "patient/patient.write",
    "patient/practitioner.read",
    "patient/practitioner.write",
    "patient/prescription.read",
    "patient/procedure.read",
    "system/allergy.read",
    "system/appointment.read",
    "system/drug.read",
    "system/facility.read",
    "system/facility.write",
    "system/immunization.read",
    "system/insurance_company.read",
    "system/insurance_company.write",
    "system/list.read",
    "system/medical_problem.read",
    "system/patient.read",
    "system/patient.write",
    "system/practitioner.read",
    "system/practitioner.write",
    "system/prescription.read",
    "system/procedure.read",
    "user/allergy.read",
    "user/appointment.read",
    "user/drug.read",
    "user/facility.read",
    "user/facility.write",
    "user/immunization.read",
    "user/insurance_company.read",
    "user/insurance_company.write",
    "user/list.read",
    "user/medical_problem.read",
    "user/patient.read",
    "user/patient.write",
    "user/practitioner.read",
    "user/practitioner.write",
    "user/prescription.read",
    "user/procedure.read"
  ],
  "response_types_supported": [
    "code",
    "token",
    "id_token",
    "code token",
    "code id_token",
    "token id_token",
    "code token id_token"
  ],
  "code_challenge_methods_supported": [
    "S256",
    "plain"
  ],
  "grant_types_supported": [
    "authorization_code",
    "password",
    "refresh_token"
  ],
  "response_modes_supported": [
    "query",
    "fragment",
    "form_post"
  ],
  "subject_types_supported": [
    "public"
  ],
  "claims_supported": [
    "profile",
    "email",
    "email_verified",
    "phone",
    "phone_verified",
    "family_name",
    "given_name",
    "fhirUser",
    "locale",
    "api:oemr",
    "api:fhir",
    "api:port",
    "api:pofh",
    "aud",
    "iat",
    "iss",
    "exp",
    "sub"
  ],
  "require_request_uri_registration": [
    "false"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_post"
  ],
  "token_endpoint_auth_signing_alg_values_supported": [
    "RS256"
  ],
  "claims_locales_supported": [
    "en-US"
  ],
  "ui_locales_supported": [
    "en-US"
  ]
}
```

sjpadgett commented:
From a request with a fhirUser and api:fhir this is the id_token claims:
```
{
  "aud": "N0qu-5BcThDbOosnXR0D066OtrsaZ7RpJCGNkbo7mk8",
  "iss": "https://opensourcedemr.us/oauth2/default",
  "iat": 1609026274,
  "exp": 1609029874,
  "sub": "9232831c-71d7-4910-b55d-103500de28b5",
  "fhirUser": true,
  "api:fhir": true,
  "email": "me@me.com",
  "email_verified": true,
  "nonce": "ab6d3ae4797dddcd01378796ef578f05"
}
```
And access_token:
```
{
  "aud": "N0qu-5BcThDbOosnXR0D066OtrsaZ7RpJCGNkbo7mk8",
  "jti": "7ff85c4f8f6c3d778bab958842a8e608f181509a148cc66225b05bb62d6891936d51e4e444a33eb6",
  "iat": 1609027397,
  "nbf": 1609027397,
  "exp": 1609030997,
  "sub": "9232831c-71d7-4910-b55d-103500de28b5",
  "scopes": [
    "openid",
    "fhirUser",
    "api:fhir",
    "email",
    "user/AllergyIntolerance.read",
    "user/CareTeam.read",
    "user/Condition.read",
    "user/Encounter.read",
    "user/Immunization.read",
    "user/Location.read",
    "user/MedicationRequest.read",
    "user/Observation.read",
    "user/Organization.read",
    "user/Organization.write",
    "user/Patient.read",
    "user/Patient.write",
    "user/Practitioner.read",
    "user/Practitioner.write",
    "user/Procedure.read",
    "nonce",
    "site:default"
  ]
}
```
{
  "aud": "N0qu-5BcThDbOosnXR0D066OtrsaZ7RpJCGNkbo7mk8",
  "jti": "7ff85c4f8f6c3d778bab958842a8e608f181509a148cc66225b05bb62d6891936d51e4e444a33eb6",
  "iat": 1609027397,
  "nbf": 1609027397,
  "exp": 1609030997,
  "sub": "9232831c-71d7-4910-b55d-103500de28b5",
  "scopes": [
    "openid",
    "fhirUser",
    "api:fhir",
    "email",
    "user/AllergyIntolerance.read",
    "user/CareTeam.read",
    "user/Condition.read",
    "user/Encounter.read",
    "user/Immunization.read",
    "user/Location.read",
    "user/MedicationRequest.read",
    "user/Observation.read",
    "user/Organization.read",
    "user/Organization.write",
    "user/Patient.read",
    "user/Patient.write",
    "user/Practitioner.read",
    "user/Practitioner.write",
    "user/Procedure.read",
    "nonce",
    "site:default"
  ]
}
![image](https://user-images.githubusercontent.com/1859985/103161262-10b2ce00-47ad-11eb-896f-8bbd61e6892a.png)

Cleaning up to push

sjpadgett commented:
@bradymiller @adunsulag 
Look over and let me know then i'll bring in.
Right now, i'm having cake and eating it too:)

adunsulag commented:
@sjpadgett Just FYI this is what I did to UserEntity->getClaims() to pass back the fhirUser parameter.  This made the fhirUser tests on inferno pass for the Launch EHR context.  I'll adjust it once you've got your centralized issuer stuff in.
```
public function getClaims()
    {
        $claimsType = ($_REQUEST['grant_type'] === 'client_credentials') ? 'client' : 'oidc';
        if ($claimsType === 'oidc') {
            $uuidToUser = new UuidUserAccount($this->identifier);
            $fhirUser = '';
            $userRole = $uuidToUser->getUserRole() == 'users';
            if ($userRole == 'users') {
                // TODO: adunsulag check with brady/sjpadget on whether this should be Practioner or Person, it has to
                // be one of those resource types and you have to be able to retrieve it via a FHIR endpoint but I'm not
                // sure a site admin is classified as a 'practioner'.
                // TODO: adunsulag we should see if there is a better way like FHIRRouteResolver for a given resource endpoint...
                $fhirUser = $GLOBALS['site_addr_oath'] . $GLOBALS['web_root'] . '/apis/' . $_SESSION['site_id'] . "/fhir/Practitioner/" . $this->identifier;
            } else if ($userRole == 'patients') {
                $fhirUser = $GLOBALS['site_addr_oath'] . $GLOBALS['web_root'] . '/apis/' . $_SESSION['site_id'] . "/fhir/Patient/" . $this->identifier;
            } else {
                SystemLogger::instance()->error("user role not supported for fhirUser claim ", ['role' => $userRole]);
            }

            SystemLogger::instance()->debug("fhirUser claim is ", ['role' => $userRole, 'fhirUser' => $fhirUser]);
```

bradymiller commented:
![2DV](https://user-images.githubusercontent.com/278968/103162989-b285eb00-47ac-11eb-9b33-cbfdc21c39c3.gif)

bradymiller commented:
btw, gonna bring in a more focused composer.lock that just updates the one package (since we are very close to releasing 6.0 just don't want any surprises :) )

bradymiller commented:
actually gonna revert my composer.lock stuff I just did since looks like league/oauth2-server also got updated and should get that into 6.0 (5 days will be enough time to ensure another updated package doesn't break something :) )

sjpadgett commented:
@bradymiller I updated league 'cause I needed to extend JWT builder for issuer so I went ahead and modified builder to be compatible with JWT lib v4 so as not to get caught when JWT v4 comes in.

@adunsulag Issuer is now set in dispatcher and auth controller as $apisBaseFullUrl(gets set in dispatch after globals.php) and $authServer->authBaseFullUrl.

sjpadgett commented:
/fhir/Practitioner/ is what is called for I believe but Person makes the most sense.
